### PR TITLE
[WIP] Update module template to make it editable in an Android app

### PR DESCRIPTION
**Pull request by stevemessick**
@xster These changes make the module generated by `flutter create -t module` editable when it is part of a larger Android app. The included script does some work to add the module to the Android app and make it runnable as add-to-app describes. We need to discuss how to implement the changes made by the script. Should it be done by the `flutter create` command? The files modified by the script are not part of the Flutter module. Another option is to document the process in detail and not implement it anywhere, or only in the Android Studio plugin.

The Android app needs to be carefully generated.
 - Min API 24
 - Use Java, not Kotlin
 - Do not use AndroidX
 
 In addition, to use the 'regen' script (below):
 - Name it ModTest
 - Put it in ~/AndroidStudioProjects
 - Use Basic Activity

Gradle sync causes the Flutter module to lose its Dart library. Dart has to be re-enabled
for the Flutter module after every sync. This happens in both IntelliJ and Android Studio.

If the selected launch config is for Android then 'Flutter Attach' is disabled in the Run
menu, along with hot reload, etc. Switching to the Flutter run config enables the menu items.

Generate an Android app named 'ModTest' using Android Studio 3.4 or 3.5. (IntelliJ generates a
project that doesn't work properly. AS 3.6 forces AndroidX; the Flutter plugin and add-to-app docs still need to be updated to work with AndroidX.)
Make sure the wizard has correct settings for the parameters
mentioned above. If it goes anywhere other than ~/AndroidStudioProjects/ModTest then make
the corresponding change to the `regen` script (see below).

Shut down the IDE.

Use the `regen` script to create a Flutter module named 'sample' in the ModTest directory.
The script edits a few files to link the module as an add-to-app module in the app.

Restart the IDE on ModTest. From this point either
IntelliJ or Android Studio can be used. In Android Studio, change the
Project drop-down from 'Android' to 'Project' in order to see Dart files.

Modify the ModTest app as described in the add-to-app documentation.
Open ModTest/sample/lib/main.dart and ensure Dart support is enabled.

Launch the Android app using the run config named 'app'.
Switch to the Flutter run config then do Run > Flutter Attach. Set a breakpoint
at the `_counter++;` line then tap the Mail icon on the Android app. The Flutter
view is displayed. Tap the '+' button to stop at the breakpoint.

The `regen` script creates a Flutter module and edits an IntelliJ configuration file
to add the module to the project. It also adds the module to the Gradle build script
and links the module's Android wrapper into the Android app.

regen script:
```
#!/bin/bash

# Script to add a Flutter module to an Android app and make it usable
# as an add-to-app module.

cd ~/AndroidStudioProjects/ModTest
rm -rf sample

flutter create --template=module --org=com.example.sample sample

cat > .idea/modules.xml <<this_is_the_end
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="ProjectModuleManager">
    <modules>
      <module fileurl="file://\$PROJECT_DIR$/ModTest.iml" filepath="\$PROJECT_DIR$/ModTest.iml" group="ModTest" />
      <module fileurl="file://\$PROJECT_DIR$/app/app.iml" filepath="\$PROJECT_DIR$/app/app.iml" group="ModTest/app" />
      <module fileurl="file://\$PROJECT_DIR$/sample/sample.iml" filepath="\$PROJECT_DIR$/sample/sample.iml" group="ModTest/sample" />
    </modules>
  </component>
</project>
this_is_the_end

#cd sample;flutter make-host-app-editable;cd ..
#echo 'include ":android"' > sample/settings.gradle

# If doing make-host-app-editable then also need to define flutter.sdk
#echo 'flutter.sdk=../src/flutter/flutter'>>sample/.android/local.properties

cat > settings.gradle <<this_is_the_end
rootProject.name='ModTest'
include ':app'
include ':sample'
setBinding(new Binding([gradle: this]))
evaluate(new File(
        settingsDir,
        'sample/.android/include_flutter.groovy'
))
this_is_the_end

# The compileOptions block is not needed if the min API is set to 26.
sed <app/build.gradle >app/temp_file \
-e '/^dependencies/a\
\    implementation project\(":flutter"\)
' \
-e '/^android/a\
\    compileOptions {\
\        sourceCompatibility 1.8\
\        targetCompatibility 1.8\
\    }
'
mv app/temp_file app/build.gradle
```
cc @devoncarew 
Edit: Fixes https://github.com/flutter/flutter/pull/24311

### CODE REVIEW

File: packages/flutter_tools/templates/module/common/build.gradle.tmpl [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Generated file. Do not edit. Makes Android Studio index the Flutter module.
+
+buildscript {}
```
Comment (xster) [inline packages/flutter_tools/templates/module/common/build.gradle.tmpl]: Ah this is how it works? Interesting.

**Comment by xster** [comment]
> The included script does some work to add the module to the Android app and make it runnable as add-to-app describes. We need to discuss how to implement the changes made by the script. 

Which scripts are you referring to? There are only iml files in this PR. 

If you meant any changes to the iml and how to migrate existing projects, I think we don't need to make it super automatic. Asking people to manually copy things is ok.

**Comment by xster** [comment]
> Use Java, not Kotlin
> Do not use AndroidX

Can you document the reasoning behind this?

**Comment by xster** [comment]
> AS 3.6 forces AndroidX; the Flutter plugin and add-to-app docs still need to be updated to work with AndroidX 

If this is the underlaying issue, @blasten is close to fixing it

**Comment by stevemessick** [comment]
> > Use Java, not Kotlin
> > Do not use AndroidX
> 
> Can you document the reasoning behind this?

It is pragmatic. The original [documentation](https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps) is written for Java and Android. It will need to be updated for Kotlin and AndroidX. Also, the plugin needs to be updated for Android X.

**Comment by xster** [comment]
@gspencergoog: the spriti of https://github.com/flutter/flutter/pull/12958 isn't to not let there be any knowledge of .iml files in users' projects, it's just to remove .iml files from the flutter/flutter repo right?

**Comment by xster** [comment]
After understanding this a bit more, I'm of the opinion that the modifications done in the script (which are purely for the sake of IntelliJ) should be done by the IntelliJ plugin. 

it seems like the main needed edits that aren't standard add-to-app procedure are the creation of the .idea/modules file and the addition of include ":sample" in the settings.gradle (not sure if this is mandatory for the scheme to work). 

if that's the case, perhaps something we can do is monitor the Gradle syncs of Android projects and do nothing by default, but once they start to add-to-app, they'll end up with :flutter:.+ tasks from the sync. At that point, we offer a notification prompt from the flutter plugin and edit files if they agree to co-edit. (this is all imagineering, I don't know if it's doable with IntelliJ plugin APIs)

wrt file structures, I think being able to let imls reference non sibling Gradle files is ideal. Otherwise, adding the child .android and synlinking back to the parent is probably the next pragmatic fix. Flipping the host/flutter file structure is probably cleaner but it's likely a multi-quarter task.

**Comment by stevemessick** [comment]
Close due to replacement by #flutter/flutter-intellij/pull/3850

**Pull request closed without merging**